### PR TITLE
Update notary change flow to support encumbrances

### PR DESCRIPTION
**Pull request by adagys**
Also moved the encumbrance pointer from ContractState to TransactionState - during the notary change tx we sometimes need to modify the encumbrance pointer to reflect the new output indexes.

Continuing from the discussion in Discourse, this is the minimal fix that makes encumbrances usable, without introducing a new type of state.

Will clean up & add tests later, just want to get some feedback first.

### CODE REVIEW

**Comment by adagys** [comment]
Primary reviewer @mikehearn

File: core/src/main/kotlin/net/corda/core/contracts/TransactionTypes.kt [hunk 1]
```diff
@@ -90,13 +90,25 @@ sealed class TransactionType {
...
             for ((i, output) in tx.outputs.withIndex()) {
-                val encumbranceIndex = output.data.encumbrance ?: continue
+                val encumbranceIndex = output.encumbrance ?: continue
                 if (encumbranceIndex == i || encumbranceIndex >= tx.outputs.size) {
                     throw TransactionVerificationException.TransactionMissingEncumbranceException(
                             tx, encumbranceIndex,
                             TransactionVerificationException.Direction.OUTPUT)
                 }
             }
+            // Verify encumbrance contracts
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/contracts/TransactionTypes.kt]: Why is this added? They are just normal states so they will have their contracts checked immediately after verifyEncumbrances returns anyway, right?

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt [hunk 2]
```diff
@@ -159,7 +159,7 @@ open class TransactionBuilder(
         return outputs.size - 1
     }
 
-    fun addOutputState(state: ContractState, notary: Party) = addOutputState(TransactionState(state, notary))
+    fun addOutputState(state: ContractState, notary: Party, encumbrance: Int? = null) = addOutputState(TransactionState(state, notary, encumbrance))
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: Could you tag this with @JvmOverloads to make the generated Java API a bit nicer.

File: core/src/main/kotlin/net/corda/flows/NotaryChangeFlow.kt [hunk 3]
```diff
@@ -36,17 +34,66 @@ object NotaryChangeFlow : AbstractStateReplacementFlow<Party>() {
...
             tx.signWith(myKey)
 
             val stx = tx.toSignedTransaction(false)
+
             return Pair(stx, participants)
         }
+
+        /**
+         * Adds the notary change state transitions to the [tx] builder for the [originalState] and its encumbrance
+         * state chain (encumbrance states migh be themselves encumbered by other states).
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/flows/NotaryChangeFlow.kt]: typo: might

File: core/src/main/kotlin/net/corda/flows/NotaryChangeFlow.kt [hunk 4]
```diff
@@ -36,17 +34,66 @@ object NotaryChangeFlow : AbstractStateReplacementFlow<Party>() {
...
+         *
+         * @return union of all added states' participants
+         */
+        private fun resolveEncumbrances(tx: TransactionBuilder): Iterable<CompositeKey> {
+            val stateRef = originalState.ref
+            val txId = stateRef.txhash
+            val issuingTx = serviceHub.storageService.validatedTransactions.getTransaction(txId)
+            val outputs = issuingTx!!.tx.outputs
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/flows/NotaryChangeFlow.kt]: Might be nice to have a better error message than an NPE here if the invariant that txid is locally stored is violated.

File: core/src/main/kotlin/net/corda/flows/NotaryChangeFlow.kt [hunk 5]
```diff
@@ -36,17 +34,66 @@ object NotaryChangeFlow : AbstractStateReplacementFlow<Party>() {
...
+            var newOutputPosition = 0
+            while (nextStateIndex != -1) {
+                val nextState = outputs[nextStateIndex]
+                tx.addInputState(StateAndRef(nextState, StateRef(txId, nextStateIndex)))
+                participants.addAll(nextState.data.participants)
+
+                if (nextState.encumbrance == null) {
+                    val modifiedState = TransactionState(nextState.data, modification)
+                    tx.addOutputState(modifiedState)
+                    nextStateIndex = -1
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/flows/NotaryChangeFlow.kt]: What's wrong with just using `break` and making it a `while (true)` loop?

File: core/src/main/kotlin/net/corda/flows/NotaryChangeFlow.kt [hunk 6]
```diff
@@ -36,17 +34,66 @@ object NotaryChangeFlow : AbstractStateReplacementFlow<Party>() {
...
+        private fun resolveEncumbrances(tx: TransactionBuilder): Iterable<CompositeKey> {
+            val stateRef = originalState.ref
+            val txId = stateRef.txhash
+            val issuingTx = serviceHub.storageService.validatedTransactions.getTransaction(txId)
+            val outputs = issuingTx!!.tx.outputs
+
+            val participants = mutableSetOf<CompositeKey>()
+
+            var nextStateIndex = stateRef.index
+            var newOutputPosition = 0
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/flows/NotaryChangeFlow.kt]: You might want to either set this to tx.outputs.size or assert on the outputs list being empty when this function is called.

File: core/src/main/kotlin/net/corda/core/contracts/TransactionTypes.kt (hunk 1 above)
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/contracts/TransactionTypes.kt]: Ah sorry, this is a leftover from my previous changes

**Comment by adagys** [comment]
Ok, back to you @mikehearn
